[DM-30063] Initial version

rra opened the pull request.

rra committed:
Unpin documenteer

rra committed:
Initial version

rra committed:
Update build system

Use the latest Sphinx configuration and GitHub Actions workflow.

rra committed:
Enable fatal warnings

Tell Sphinx to exit with an error if there are any warnings.  Fix
various warnings that uncovered.

rra committed:
Add dependabot configuration

rra committed:
Fix minor errors

Fix minor markup errors and one completed task, found during
proofreading.

rra committed:
Fix reference to DMTN-193

DMTN-193 is the relevant document, not DMTN-163.

rra committed:
Add implementation details for GitHub org membership

rra committed:
Update abstract

Use the same introductory wording as DMTN-234.

rra committed:
Add note for document series

Add a note explaining how this tech note fits into the larger
series of tech notes about the identity management system.

rra committed:
Update catalog of tech notes

Add additional tech notes that are part of the identity management
series or cover related topics.

rra committed:
Flesh out the implementation overview

Delete stuff covered in DMTN-234, explain the components, and add
a diagram.

rra committed:
Further work on architecture diagrams

Add a more detailed diagram of the general access case.  Include
the Gafaelfawr UI in the Science Platform diagram and make a few
things more consistent.

rra committed:
Document Notebook Aspect integration

Document how Gafaelfawr is integrated into the Notebook Aspect,
including a sequence diagram of the interaction.

rra committed:
Flesh out identity management section

Document the implementation of COmanage and GitHub (the latter
mostly about how group names are contructed).

rra committed:
Merge information from several sections

Merge all the GitHub data together, except for the operational note
about releasing organizational information, which is moving to the
Phalanx documentation.  Merge the federated identities section into
the identity management section.

rra committed:
Add browser flows with sequence diagrams

Import the browser authentication flow information from the
Gafaelfawr docs and make three sequence diagrams for each identity
provider.  Flesh out the details of specific behavior of each
identity provider.

rra committed:
Add documentation of storage formats

Much of this is copied from SQR-049, but some of that document
was out of date, and it didn't document the cookie format or
the Firestore database.

rra committed:
Add documentation of token flows

rra committed:
Document bootstrap username, Portal Aspect

Note the name of the user (for logging purposes) of the bootstrap
token.  Document special authentication handling for the Portal
Aspect.

rra committed:
Document caching behavior

Also include the locking around the caching, and the strategy for
reusing existing internal and notebook tokens.

rra committed:
Add documentation of OpenID Connect support

Document the use of Gafaelfawr as an OpenID Connect provider.

rra committed:
Add token API implementation details

Mostly from SQR-049, with implementation decisions removed and
omitting pieces that have not yet been implemented.

rra committed:
Move remaining work to SQR-069

Consolidate all the remaining work sections in the document for
history and discussion.

rra committed:
Flesh out token UI section

rra committed:
Add a pointer to SQR-069 for remaining work

rra committed:
Document the GafaelfawrServiceToken resource

rra committed:
Clarify lack of scope checks on OIDC auth

rra committed:
Spell check

rra committed:
Add discussion of NetworkPolicy

rra committed:
Don't over-link to Gafaelfawr

No need to make every mention a link.

rra committed:
Capitalize LDAP in flow diagrams

rra committed:
Tweak References section

Clarify the role of SQR-049 now, and explicitly say that this
document is the primary implementation document.

rra committed:
Add a section on logging

Import some of the implementation details of logging from the
Gafaelfawr documentation, and move the specific services section
to a more natural location.

rra committed:
Clarify role of SQR-039

rra committed:
Add more information about OIDC limitations

rra committed:
Specify the LDAP attributes used with COmanage

Rather than speaking generically about user identity information,
say specifically that we use displayName and mail.

rra committed:
Add link to onboarding templates

The pages shown during COmanage onboarding are controlled by
templates in GitHub.  Link to that repository.

rra committed:
Move SQR-049 and SQR-069 in references

Move SQR-049 to historic and flesh out the reasons why it may
still be of interest.  Move SQR-069 to an implementation tech note
instead of marking it as historic.

rra committed:
Document special JavaScript unauthenticated handling

Document the special checks used to return 403 rather than 401 to
JavaScript requests from unauthenticated users.

rra committed:
Document uncached 403 responses

Document the workaround to force NGINX to return uncached 403
responses.

rra committed:
Add reference to DMTN-169 (Butler access control)

rra committed:
Add reference to DMTN-182

This document covers Butler's proposed use of groups for access
control.

rra committed:
Fix typo in logging section

rra committed:
Minor clarifications for OpenID codes

Mention that OpenID codes issued by Gafaelfawr are deleted as soon
as they are redeemed.

rra committed:
Add more details on tokens and JWTs

Incorporate some details that were in the Gafaelfawr documentation.

rra committed:
Document /logout and redirect restrictions

More information taken from the Gafaelfawr documentation.  Also
add PUT to the list of HTTP verbs requiring a CSRF header.

rra committed:
Explain data precedence for user information

frossie commented on index.rst:
```diff
@@ -1,61 +1,1351 @@
-  Subsections
-  -----------
+   This is part of a tech note series on identity management for the Rubin Science Platform.
+   The other primary documents are DMTN-234_, which describes the high-level design; and SQR-069_, which provides a history and analysis of the decisions underlying the design and implementation.
```
The DMTN-234 link is pointing to DMTN-225

frossie commented on index.rst:
```diff
@@ -1,61 +1,1351 @@
+With respect to the identity management system, these differ primarily in the choice of the first two components.
 
-.. sectnum::
+General access deployments of the Science Platform use CILogon_ as the source of user authentication, and COmanage_ as the repository of identity information.
```
Passive voice makes this sound inevitable, maybe "We have chosen to [] for our main general access deployment.

frossie commented on index.rst:
```diff
@@ -1,61 +1,1351 @@
+   Detail of the components for identity management for a general access deployment of the Science Platform.
+   The Science Platform aspects and services are represented here by a single service to make the diagram simpler.
+
+Restricted access deployments use either GitHub or a local `OpenID Connect`_ authentication provider as the source of user authentication; and one of GitHub, a local LDAP server, or the OpenID Connect authentication provider as the source of identity information.
```
```suggestion
We also support more restricted access deployments that use either GitHub or a local `OpenID Connect`_ authentication provider as the source of user authentication; and one of GitHub, a local LDAP server, or the OpenID Connect authentication provider as the source of identity information.
```

frossie commented on index.rst:
```diff
@@ -1,61 +1,1351 @@
+
+The identity management component of the system authenticates the user and maps that authentication to identity information.
+
+In general access deployments, authentication is done via the OpenID Connect protocol using CILogon, which in turn supports SAML and identity federations as well as other identity providers such as GitHub and Google.
```
Instead of differentiation on general or restricted access, just differentiate on the provider choice. "In our federated identity deployments" or "in our CILogon deployments" etc. There's no point baking in the general/restricted differentiation in a technology note

frossie reviewed (approved):
Very thorough. I would just just be very upfront (maybe even in the intro) and go "We currently support three sources of identity:
CILogon [section blah]
Github [section blah]
OpenID [section blah]

in a top-level structure that makes those turn up on the sidebar. This is because even though this document is very significant to us, the most likely extenal reader is one who is trying to chose how to hook up their deployment, or checking what methods are supported and how (Stelios usecase or similar)

rra committed:
Drop general and restricted access terminology

This was confusing and was unnecessary abstraction.  Just call them
federated identity, GitHub, and local OpenID Connect deployments
following the current wording in DMTN-234, and adjust the wording
accordingly.

rra committed:
Update for review feedback

Fix a few remaining wording issues and one incorrect link.
